core/market/src/protocol/negotiation/messages.rs

```diff
@@ -130,7 +131,12 @@ impl RpcMessage for AgreementReceived {
 #[serde(rename_all = "camelCase")]
 pub struct AgreementApproved {
     pub agreement_id: AgreementId,
-    // TODO: We should send here signature.
+    /// This will be placed in `approved_signature` Agreement field.
+    pub signature: String,
+    /// This timestamp will differ from timestamp, when Agreement will be updated in
+    /// database to `Approved` state and `AgreementApprovedEvent` timestamp either.
+    /// But we can't set it to any value, because we should include this field in signature.
```

Comment:
I don't know how to clarify this, but I have feeling, that you missed the main point, that requires some consideration:
We have `approved_ts`  in Agreement, but we can't set this value at the moment, when Agreement changed state to `Approved`. And we can't do this, because this timestamp should be signed by Provider and he sends his signature, when Proposal changes state to `Approving`.

Reply:
You're right: the `approved_ts` should be signed by both sides, so it needs to be specified when the provider sends this message, with its only signature for the Agreement. Then Requestor will create its second signature which will also cover the `approved_ts`.

My point is that the last sentence of your comment is not clear, or even false. We can and must set the `approved_ts` here. And you are doing this in the code below:

https://github.com/golemfactory/yagna/pull/1015/files#diff-b5dd89d987b0df3b7532481d3ef3990172d23e8a6b746e625ff346ee6ffa9ac0R125-R127